src/mono/wasm/test-main.js

```diff
@@ -243,53 +242,95 @@ const App = {
 };
 globalThis.App = App; // Necessary as System.Runtime.InteropServices.JavaScript.Tests.MarshalTests (among others) call the App.call_test_method directly
 
+function configureRuntime(dotnet, runArgs, INTERNAL) {
+    dotnet
+        .withVirtualWorkingDirectory(runArgs.workingDirectory)
+        .withEnvironmentVariables(runArgs.environmentVariables)
+        .withDiagnosticTracing(runArgs.diagnosticTracing)
+        .withDiagnosticTracing(true)
+        .withExitOnUnhandledError()
+        .withExitCodeLogging()
+        .withElementOnExit();
+
+    if (is_node) {
+        dotnet
+            .withEnvironmentVariable("NodeJSPlatform", process.platform)
+            .withAsyncFlushOnExit();
+
+        const modulesToLoad = runArgs.environmentVariables["NPM_MODULES"];
+        if (modulesToLoad) {
+            dotnet.withModuleConfig({
+                onConfigLoaded: (config) => {
+                    loadNodeModules(config, INTERNAL.require, modulesToLoad)
+                }
+            })
+        }
+    }
+    if (is_browser) {
+        dotnet.withEnvironmentVariable("IsWebSocketSupported", "true");
+    }
+    if (runArgs.runtimeArgs.length > 0) {
+        dotnet.withRuntimeOptions(runArgs.runtimeArgs);
+    }
+    if (runArgs.debugging) {
+        dotnet.withDebugging(-1);
+        dotnet.withWaitingForDebugger(-1);
+    }
+    if (runArgs.forwardConsole) {
+        dotnet.withConsoleForwarding();
+    }
+}
+
+async function dry_run(runArgs) {
+    try {
+        console.log("Silently starting separate runtime instance as another ES6 module to populate caches...");
+        // this separate instance of the ES6 module, in which we just populate the caches
+        const { dotnet, exit, INTERNAL } = await loadDotnet('./dotnet.js?dry_run=true');
+        mono_exit = exit;
+        configureRuntime(dotnet, runArgs, INTERNAL);
+        // silent minimal startup
+        await dotnet.withConfig({
+            forwardConsoleLogsToWS: false,
+            diagnosticTracing: false,
+            appendElementOnExit: false,
+            logExitCode: false,
+            pthreadPoolSize: 0,
+            exitAfterSnapshot: true
+        }).create();
+    } catch (err) {
+        if (err && err.status !== 0) {
+            return false;
+        }
+    }
+    console.log("Separate runtime instance finished loading.");
+    return true;
+}
+
 async function run() {
     try {
+        const runArgs = await getArgs();
+        console.log("Application arguments: " + runArgs.applicationArguments.join(' '));
+
+        if (is_browser) {
+            const dryOk = await dry_run(runArgs);
```

Comment:
Just to verify: This will *always* create a snapshot right? It won't reuse a matching one it happens to find in the browser cache if (for example) someone is re-running tests in a browser tab with the debugger open to troubleshoot?

To me `exitAfterSnapshot: true` implies 'always make a new snapshot and save it' but I'm not certain from reading everything whether that's the intent and behavior here.

Reply:
I added comment.